File: day25/day25.py
from typing import List, Optional


class Blizzard:
    def getInput(self) -> List[str]:
        inputFile = "input-test.txt" if self.useTest else "input.txt"
        data = []
        with open(inputFile, "r") as file1:
            for line in file1.readlines():
                data.append(list(line.strip()))
            return data

    def __init__(self, useTest: Optional[bool] = False) -> None:
        self.useTest = useTest
        self.map = self.getInput()


if __name__ == "__main__":
    blizzard = Blizzard(False)
    print("Day 25 part 1:")
    print("Day 25 part 2:")

File: 2024/day25/day25.js
const fs = require("fs");

class Code {
  getInput() {
    const inputFile = this.useTest ? "input-test.txt" : "input.txt";
    try {
      const data = fs.readFileSync(inputFile, "utf8").trim().split(/\r?\n/);
      const schemaList = [];
      let schema = [];
      data.forEach((line) => {
        if (!line) {
          schemaList.push(schema);
          schema = [];
          return;
        }
        schema.push(line);
      });
      schemaList.push(schema);
      return schemaList;
    } catch (err) {
      throw err;
    }
  }

  constructor(useTest = false) {
    this.useTest = useTest;
    this.schemaList = this.getInput();
    this.splitKeys();
  }

  splitKeys() {
    this.keys = [];
    this.locks = [];

    for (const schema of this.schemaList) {
      const counts = Array.from({ length: schema[0].length }, () => -1);
      for (const line of schema) {
        for (let i = 0; i < line.length; i++) {
          if (line[i] === "#") counts[i]++;
        }
      }
      if (schema[0] === "#####") this.locks.push(counts);
      else this.keys.push(counts);
    }
  }

  isValid(key, lock) {
    for (let i = 0; i < key.length; i++) {
      if (key[i] + lock[i] > 5) return false;
    }
    return true;
  }

  getCombo() {
    let valid = 0;
    for (const key of this.keys) {
      for (const lock of this.locks) {
        if (this.isValid(key, lock)) valid++;
      }
    }
    return valid;
  }
}

const code = new Code();
console.log("Day 25 part 1:", code.getCombo());
